infrastructure/application_stack.py

```diff
@@ -103,3 +105,5 @@ def __init__(self, scope: constructs.Construct, stack_id: str) -> None:
             OpenTopography(
                 self, "opentopography", env_name=env_name, storage_bucket=storage.storage_bucket
             )
+
+        atexit.register(delete_packaging_dir)
```

Comment:
Is it possible to do this within `bundled_code`? It would be nice to keep it together with the code which creates it. I'm not sure whether registering it more than once would be a problem, though.

Reply:
This isn't possible as far as I can tell. `bundled_code` returns `aws_lambda.Code.from_asset(path=".", bundling=bundling_options)` where `bundling_options` is called. We would prematurely delete `PACKAGING_DIR` before it gets called / used.  

We also package lambda at a several places. Cleaning up at `bundled_code` would mean we recreate `PACKAGING_DIR` again over and over for each lambda?

Can you think of a more elegant way to approach this?

Reply:
> This isn't possible as far as I can tell. `bundled_code` returns `aws_lambda.Code.from_asset(path=".", bundling=bundling_options)` where `bundling_options` is called. We would prematurely delete `PACKAGING_DIR` before it gets called / used.

Based on a quick scan of [the docs](https://docs.python.org/3/library/atexit.html), I don't think that's correct:

>  Functions thus registered are automatically executed upon normal interpreter termination.

So it should only be run once the overarching Python process finishes. Which, as far as I can tell, should result in the same behaviour as the original code (with the exception of `rmtree` running multiple times for the same directory).

Reply:
Interesting. This does work, it seems. Although it will throw a bunch of errors like so:

```Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib/python3.9/shutil.py", line 722, in rmtree
    orig_st = os.lstat(path)
FileNotFoundError: [Errno 2] No such file or directory: 'geostore/.lambda_out_5y5bg5o9'
```

due to `rmtree` being called several times over and over.

Bit of a tradeoff in terms between code tidiness vs error thrown (which could be hard to trace in the future)?

If we  only call `atexit.register(lambda: shutil.rmtree(PACKAGING_DIR))` once in `application_stack.py`, we could put a short one line comment like `# cleaning up lambda packaging directory in bundled_code`, where it would be obvious, vs trying to track down `FileNotFoundError: [Errno 2] No such file or directory: 'geostore/.lambda_out_5y5bg5o9'` for each deployment. This could be confusing for anyone new coming into the code?

Reply:
We could use [`ignore_errors=True`](https://docs.python.org/3/library/shutil.html#shutil.rmtree), although that would also silence other errors. So yeah, maybe a single execution is the way to go. Up to you.

Reply:
> Actually, could we register the handler where we create the packaging directory?

Good point.